src/serde.rs

```diff
@@ -20,13 +20,63 @@ impl Serialize for Ulid {
     }
 }
 
+struct UlidVisitor(&'static str);
+impl<'de> serde::de::Visitor<'de> for UlidVisitor {
+    type Value = Ulid;
+
+    fn expecting(&self, formatter: &mut core::fmt::Formatter) -> core::fmt::Result {
+        formatter.write_str(self.0)
+    }
+
+    fn visit_str<E>(self, value: &str) -> Result<Self::Value, E>
+    where
+        E: serde::de::Error,
+    {
+        #[cfg(feature = "uuid")]
+        if matches!(value.len(), 32 | 36 | 38 | 45) {
+            return match uuid::Uuid::try_parse(value) {
+                Ok(a) => Ok(Ulid::from(a)),
+                Err(e) => Err(serde::de::Error::custom(e)),
+            };
+        }
+        Ulid::from_string(value).map_err(serde::de::Error::custom)
+    }
+
+    fn visit_bytes<E>(self, v: &[u8]) -> Result<Self::Value, E>
+    where
+        E: serde::de::Error,
+    {
+        // allow either the 16 bytes as is, or a wrongly typed str
+        match v.len() {
+            16 => {
+                let ptr = v.as_ptr() as *const [u8; 16];
+                Ok(Ulid::from_bytes(*unsafe { &*ptr }))
+            }
+            crate::ULID_LEN => Ulid::from_string(unsafe { core::str::from_utf8_unchecked(v) })
```

Comment:
Why is the bytes in v assumed to be valid utf8?

Reply:
It might be somewhat of an over-the-top optimization, but `from_string(v: &str)` only interacts with the string variable using `v.len()` and `v.as_bytes()` and doesn't actually use characters or char boundaries in any way. A better solution that still doesn't unnecessarily validate the string twice would have been to change  `base32::decode(_: &str)` to `base32::decode(_: &[u8])` and call that directly instead.

Reply:
Ah, i see, so because of like 79 `deserializer.deserialize_str(UlidVisitor("an ulid string or value"))` it forces the uft-8 to be valid. Understood thanks.